Enyo 1181 Add new layout GridListItem to sample

DavidUM opened the pull request:
## Issue

There is UX requirement for horizontal layout GridListImageItem.
## Fix

Instead of add new kind to moonstone control, just provide more samples

Enyo-DCO-1.1-Signed-off-by: David Um david.um@lge.com

Commit:
Enyo-1181 Bind selected property from model and control

Enyo-DCO-1.1-Signed-off-by: David Um david.um@lge.com

Commit:
Enyo-1181 Add 2 more gridListItem sample with multi selection support.

Enyo-DCO-1.1-Signed-off-by: David Um david.um@lge.com

Commit:
Enyo-1181 Add css file for added sample.

Enyo-DCO-1.1-Signed-off-by: David Um david.um@lge.com

ryanjduffy commented on src/moonstone-samples/lib/DataGridListSample.js:
```diff
@@ -110,12 +147,72 @@ module.exports = kind({
+
+		var moreProps = {};
+
+		switch (itemType) {
```
I'd favor if/else block here rather than switch for only 2 options

DavidUM replied:
It was supposed to have more options than just 2, but I could not make it.
How about leave switch statement for further works?
When app developers have different GUI, I'll add more option for their reference.

ryanjduffy commented on src/moonstone-samples/lib/DataGridListSample.js:
```diff
@@ -110,12 +147,72 @@ module.exports = kind({
+				}]
+			};
+			break;
+		default:
```
No need for an empty default in a switch

DavidUM replied:
It has some code but I removed it. I forgot to remove default. Let me make it. Thanks. :+1:

ryanjduffy commented on src/moonstone-samples/lib/DataGridListSample.js:
```diff
@@ -110,12 +147,72 @@ module.exports = kind({
+			this.$.gridList.destroy();
+		}
+
+		var moreProps = {};
```
Might be simpler to clone `dataListDefaults` here and then modify that copy in the conditional block below. Results in fewer objects created as well and while that's not critical for a sample, it does demonstrate better practices for anyone that might base their production code off of the sample.

DavidUM replied:
Good point. I didn't consider that. Thank you for your point.

ryanjduffy commented on src/moonstone-samples/lib/DataGridListSample.css:
```diff
@@ -0,0 +1,25 @@
+.moon-gridlist-imageitem.horizontal-gridList-image-item > .sub-caption {
+	position: relative;
+	top: 18%;
+    width: auto;
```
There's some mixed tab and space indentation here.

ryanjduffy commented on src/moonstone-samples/lib/DataGridListSample.js:
```diff
@@ -110,12 +147,72 @@ module.exports = kind({
+	itemTypeChanged: function(inSender, inEvent) {
+		this.generateDataGridList(inSender.active.value);
+	},
+	selectionTypeChanged: function(inSender, inEvent) {
```
Be sure to leave a space after `function` for all function definitions and to not prefix arguments with `in`

`selectionTypeChanged: function (sender, event) {`

ryanjduffy commented on src/moonstone-samples/lib/DataGridListSample.js:
```diff
@@ -28,9 +29,50 @@ var GridSampleItem = kind({
+
+	bindings: [
+		{from: 'caption', to: '$.caption.content'},
+		{from: 'caption', to: '$.caption.showing', kind: 'enyo.EmptyBinding'},
```
EmptyBinding needs to be `require()`d rather than referenced by string

DavidUM replied:
Oh, I missed that. Thanks.

Commit:
Enyo-1181 fix some minor fault.

Enyo-DCO-1.1-Signed-off-by: David Um david.um@lge.com

Commit:
Enyo-1181 Fix more minor fault and Reduce unnecessary object creation

Enyo-DCO-1.1-Signed-off-by: David Um david.um@lge.com

ryanjduffy commented on src/moonstone-samples/lib/DataGridListSample.js:
```diff
@@ -110,12 +148,70 @@ module.exports = kind({
+		}
+
+		createComponent = this.$.listPanel.createComponent(props, {owner: this});
+		createComponent.render();
```
One last thing ... your local variable is `createdComponent` but you're assigning to `createComponent`. I'll push a fix for this.

Commit:
fix typo

Enyo-DCO-1.1-Signed-off-by: Ryan Duffy (ryan.duffy@lge.com)

aarontam commented on src/moonstone-samples/lib/DataGridListSample.js:
```diff
@@ -79,24 +124,17 @@ module.exports = kind({
 		]}
 	],
 	bindings: [
-		{from: '.collection', to: '.$.dataList.collection'},
```
Hi @DavidUM, it seems like this binding might have been mistakenly removed? This is resulting in the "Popup List" `moon.ContextualPopup` being empty.

aarontam commented on src/moonstone-samples/lib/DataGridListSample.js:
```diff
@@ -40,26 +83,28 @@ module.exports = kind({
+			{kind: ContextualPopupDecorator, components: [
+				{kind: ContextualPopupButton, content:'Item Type'},
+				{kind: ContextualPopup, classes:'moon-6h', components: [
+					{kind: RadioItemGroup, name: 'itemTypeGroup', onActiveChanged: 'itemTypeChanged', components: [
```
The `onActiveChanged` event is being fired when this popup is first opened, causing an undesired regeneration of the DataGridList.

aarontam replied:
The `onActiveChanged` event is being fired when this popup is first opened, causing an undesired regeneration of the DataGridList. Seems like this is a pre-existing behavior with `moon.SelectableItem`, as `updateActiveValue` is called in `rendered`, which doesn't occur until the popup is actually opened (rendering the item) in this case.

@germboy (because of your history making some related tweaks in https://github.com/enyojs/moonstone/commit/77f8c73fd1c06b10fe57ed6b257b2f736c53d668) @ryanjduffy any thoughts on this and/or are there issues you can think of if we instead call `selectedChanged`(instead of `updateSelectedValue`) in `create` and remove the call in `rendered`? I realize that this can definitely be outside the scope of this PR, but if we wanted to leave the existing behavior of `moon.SelectableItem`, I think we should probably make a tweak to the sample to prevent the observed behavior.

DavidUM replied:
@aarontam and I decided to make another jira ticket to resolve this behavior.

aarontam replied:
After some further discussion, we want to fully resolve the issue here before integrating this change. So @germboy @ryanjduffy, please see the first comment when you have a chance, thanks!

ryanjduffy replied:
Pushed a fix in the sample for this behavior 93d3444fa29999b97986fe739cbf5d7b48b2e530

aarontam replied:
:+1:

Commit:
Enyo-1181 Restore binding to dataList collection to show items in popup

Enyo-DCO-1.1-Signed-off-by: David Um david.um@lge.com

Commit:
Merge branch 'Enyo-1181-davidum-2.6.0-dev' of https://github.com/enyojs/enyo-strawman into Enyo-1181-davidum-2.6.0-dev

Commit:
ensure grid list is only rerendered when the kind changes

Enyo-DCO-1.1-Signed-off-by: Ryan Duffy (ryan.duffy@lge.com)

aarontam commented on src/moonstone-samples/lib/DataGridListSample.js:
```diff
@@ -110,12 +151,70 @@ module.exports = kind({
+			this.$.gridList.destroy();
+		}
+
+		var props = utils.clone(this.dataListDefaults),
```
Hi @DavidUM, this is not performing a deep-clone, so when we modify a nested property of `props` i.e. in the case of "HorizontalGridListItem" setting the bindings array, we are changing the bindings of "GridSampleItem". Switching from "HorizontalGridListItem" to "GridSampleItem" results in the `source` binding being effectively removed, so no image shows up. I think we'll have to rework how we're generating the item properties.

ryanjduffy replied:
Simplest solution here is to drop `dataListDefaults` and assign the object literal to props. There's really no need for it to be a class member and is only used in this method.

Commit:
move data grid list item type props to local variable

utils.clone() was producing a shallow copy causing issues when
switching between item types. Moving to a local variable removes the
need for the clone.

Issue: ENYO-1181
Enyo-DCO-1.1-Signed-off-by: Ryan Duffy (ryan.duffy@lge.com)

aarontam committed:
Merge branch 'master' into Enyo-1181-davidum-2.6.0-dev

aarontam committed:
ENYO-1181: Restore custom selection overlay icon positioning.

Enyo-DCO-1.1-Signed-off-by: Aaron Tam <aaron.tam@lge.com>
